TUM-Dev/NavigaTUM
# [Search-Refactoring-1] Migrated to the Meilisearch SDK

Fixes #595
Fixes a connection bug observed in #576 (headers being parsed weird)

Fixes some combination of inputs crashing the thread (DDOS Possibility). One such search term is `%1A`, i.e. the [SUB command](https://www.w3schools.com/tags/ref_urlencode.ASP) 

## Proposed Changes (include Screenshots if possible)

- This PR exchanges to the Meilisearch SDK
- Inadvertedly some parameters are handled differently (e.g. `a;`) Given that these dont come from the ascii+numbers+common punctuation spectrum, I am not particularly concerned. This is most likely just a bit different behavour in URL parsing/encoding.
- now the calculation of `estimated_total_hits` for buildings is based upon the buildings and not on the `facetDistribution` of the merged search

## How to test this PR

1. Make searches
2. Compare them with the previous results
3. See that nothing changed

## How has this been tested?

<details><summary>By comparing the output of the previous API to the current one</summary>
<p>

```py
import itertools
import json

import requests
import string
import random
import difflib
import pprint

from tqdm import tqdm
from tqdm.contrib.concurrent import thread_map


def compare_dicts(d1, d2):
    return ('\n' + '\n'.join(difflib.ndiff(
        pprint.pformat(d1).splitlines(),
        pprint.pformat(d2).splitlines())))


URL1 = 'http://localhost:8080'  # new sdk
URL2 = 'http://localhost:8081'  # old version


def download(base: str, query: str) -> list[dict]:
    url = base + '/api/search?q=' + query
    with requests.get(url) as request:
        if request.status_code != 200:
            raise ValueError(f"Error: >{url}< returned status code {request.status_code}")
        sections: list[dict] = json.loads(request.text)["sections"]
    for s in sections:
        s["estimatedTotalHits"] = 0
    return sections


def compare(query: list[str]):
    query: str = ''.join(query)
    r1 = download(URL1, query)
    r2 = download(URL2, query)
    if r1 != r2:
        print(compare_dicts(r1, r2))
        print("===r1===", r1)
        print("===r2===", r2)
        raise ValueError(f"'{URL1}/api/search?q={query}' and '{URL2}/api/search?q={query}' returned different results for "{query}"')


def random_string(length):
    return (random.choice(BASE_CHARSET) for _ in range(length))


MAX_WORKERS = 2
BASE_CHARSET = string.printable
if __name__ == '__main__':
    for i in range(4, 4):
        thread_map(compare, itertools.permutations(BASE_CHARSET, i), desc=f"Searching with lengh {i} extensively", max_workers=MAX_WORKERS)

    random_guesses=1000
    for _ in range(30):
        permutations = [random_string(random.randint(4,30)) for _ in range(random_guesses)]
        thread_map(compare, permutations, desc=f"Searching with lengh randomly", max_workers=MAX_WORKERS)
```

</p>
</details> 


## Checklist:

- [x] This PR is Split up into atomic units making review eazy
- [x] figured out if filtering works as it should
- [x] figgure out if postprossesing of `facetDistribution` is handled correctly
